### PR TITLE
Expose shard changes action as a rest api

### PR DESCRIPTION
**Pull request by salvatore-campagna**
This PR adds a new REST API `/{index}/ccr/shard_changes` to retrieve shard-level changes, including translog operations, mapping versions, and sequence numbers. It is required for a new set of Rally benchmarks, which will be used to evaluate the impact of synthetic source on recovery time. The API accepts parameters like `from_seq_no`, `max_batch_size`, `poll_timeout`, and `max_operations_count` and is exposed only in snapshot builds.

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java [hunk 1]
```diff
@@ -272,25 +276,24 @@ public List<RestHandler> getRestHandlers(
...
-            // auto-follow APIs
-            new RestDeleteAutoFollowPatternAction(),
-            new RestPutAutoFollowPatternAction(),
-            new RestGetAutoFollowPatternAction(),
-            new RestPauseAutoFollowPatternAction(),
-            new RestResumeAutoFollowPatternAction(),
-            // forget follower API
-            new RestForgetFollowerAction()
-        );
+        return Stream.concat(
```
Comment (salvatore-campagna) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java:279]: `Arrays.asList` returns a fixed-size list. Ass a result, adding later the additional handler based on the value of `Build.current().isSnapshot()` might fail with an `UnsupportedOperationException`.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestShardChangesAction.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+     *
+     * @param response The ShardChangesAction response containing the operations to be converted.
+     * @param builder The XContentBuilder to which the converted operations will be added.
+     * @throws IOException If an error occurs while writing to the XContentBuilder.
+     */
+    private static void operationsToXContent(final ShardChangesAction.Response response, final XContentBuilder builder) throws IOException {
+        builder.field("number_of_operations", response.getOperations().length);
+        builder.field("operations");
+        builder.startArray();
+        for (final Translog.Operation operation : response.getOperations()) {
```
Comment (salvatore-campagna) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestShardChangesAction.java:217]: Maybe we can avoid this...as serializing each operation for large requests might be way too much?

Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestShardChangesAction.java:217]: I see that you already left out the actual content of the write operations, which is good for the use case this rest api is going to be used. Let's for now keep the response as is here. If this turns out the be problematic we can always drop the operations array from the response.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java (hunk 1 above)
Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java:279]: Maybe just create an array list with an if statement for the shard changes action? Looks simpler to me than all this stream stuff.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestShardChangesAction.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+
+/**
+ * A REST handler that retrieves shard changes in a specific index whose name is provided as a parameter.
```
Comment (martijnvg) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestShardChangesAction.java:45]: Maybe at the top of the java docs, mention that this rest action is only loaded in snapshot builds and its use is for benchmarking purposes only?

File: x-pack/plugin/ccr/src/javaRestTest/java/org/elasticsearch/xpack/ccr/rest/ShardChangesRestIT.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                 }
+               }
+            """;
+        createIndex(indexName, settings, mappings);
+        assertTrue(indexExists(indexName));
+
+        assertOK(client().performRequest(bulkRequest(indexName, randomIntBetween(10, 20))));
+
+        final Request shardChangesRequest = new Request("GET", shardChangesEndpoint(indexName));
+        assertOK(client().performRequest(shardChangesRequest));
```
Comment (martijnvg) [inline x-pack/plugin/ccr/src/javaRestTest/java/org/elasticsearch/xpack/ccr/rest/ShardChangesRestIT.java:86]: maybe assert the response?

**Review by martijnvg — APPROVED** [review]
Left a few nits. LGTM 👍

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestShardChangesAction.java (hunk 3 above)
Comment (salvatore-campagna) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestShardChangesAction.java:45]: I added a **note** below in the Javadoc.

File: x-pack/plugin/ccr/src/javaRestTest/java/org/elasticsearch/xpack/ccr/rest/ShardChangesRestIT.java (hunk 4 above)
Comment (salvatore-campagna) [reply x-pack/plugin/ccr/src/javaRestTest/java/org/elasticsearch/xpack/ccr/rest/ShardChangesRestIT.java]: Done

**Pull request merged**